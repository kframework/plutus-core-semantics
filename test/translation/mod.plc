(module Foo (import) (export () ())
  (define mod (lam x (lam y (builtin remainderInteger x y))))
)