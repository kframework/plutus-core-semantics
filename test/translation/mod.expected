<T>
  <plutus>
    .
  </plutus>
  <iele>
    .TopLevelDefinitions
  </iele>
  <ieleContract>
    contract Foo {
      define @mod ( %x , %y , .LocalNames ) {
        %ret = mod %x , %y
        ret %ret , .Operands
        .Instructions .LabeledBlocks
      }
      
      define @dispatch ( %closure , %param , .LocalNames ) {
        %closureId = load %closure
        %cond = cmp eq %closureId , 1
        br %cond , call0
        %ret , .LValues = call @appendArg ( %closure , %closureId , %param , .Operands )
        ret %ret , .Operands

        call0:
        %closure = add %closure , 1
        %param1 = load %closure
        %ret , .LValues = call @mod ( %param1 , %param , .Operands )
        ret %ret , .Operands
        .Instructions .LabeledBlocks
      }
      
      define @alloc ( %size , .LocalNames ) {
        %alloc_counter_ptr = 0
        %alloced = load %alloc_counter_ptr
        %isInitialized = cmp gt 0 , %alloced
        br %isInitialized , initialized
        %alloced = 1

        initialized:
        %ret = %alloced
        %alloced = add %size , %alloced
        store %alloced , %alloc_counter_ptr
        ret %ret , .Operands
        .Instructions .LabeledBlocks
      }
      
      define @appendArg ( %closure , %closureId , %param , .LocalNames ) {
        %numParams = load %closure , 1 , 1
        %numParams = add %numParams , 1
        %spaceNeeded = add %numParams , 1
        %ret , .LValues = call @alloc ( %spaceNeeded , .Operands )

        // "Location in returned closure to write to"
        %writeAddr = %ret

        // "Copy closure ID"
        store %closureId , %writeAddr

        // "Update number of parameters"
        store %numParams , %writeAddr , 1 , 1

        // "Copy arguments one at a time"

        while:
        %numParams = sub %numParams , 1
        %cond = cmp eq %numParams , 0
        br %cond , done
        %writeAddr = add %writeAddr , 1
        %closure = add %closure , 1
        %v = load %closure
        store %v , %writeAddr
        br while

        done:

        // "Add new parameter"
        %writeAddr = add %writeAddr , 1
        store %param , %writeAddr
        ret %ret , .Operands
        .Instructions .LabeledBlocks
      }
      
      .TopLevelDefinitions
    }
  </ieleContract>
  <totalFunctions>
    2
  </totalFunctions>
  <totalReadyForDispatch>
    2
  </totalReadyForDispatch>
  <dispatchedFunctionsLeft>
    0
  </dispatchedFunctionsLeft>
  <exports>
    .QualCList
  </exports>
  <functions>
    .FunctionCellSet
  </functions>
</T>
